### ISSUE TITLE
InexactError: Int64(1.0) on macOS (CI)

### ISSUE THREAD

**Issue by lmiq** [JuliaLang/julia#55878]
I experienced this (non-reproducible error) in a CI run on the `pre` release (1.11 now):

```julia
 ERROR: LoadError: TaskFailedException

    nested task error: InexactError: Int64(1.0)
```

The line where the error occurred was:

```julia
            index = floor(Int, xi) + 1
```

in which `xi` is a `Float64`. As far as I understand, the result of `floor` should always return an exactly representable Int, so the error should not occur. 

When reporting this on Zulip, it, @jlumpe (sorry if the tag is wrong), reported having observed two times similar errors:

https://github.com/mikeingold/MeshIntegrals.jl/actions/runs/10965929580/job/30452792024?pr=73#step:7:191
https://github.com/JoshuaLampert/DispersiveShallowWater.jl/actions/runs/10811172112/job/29990016675?pr=150#step:7:8511

But these with 1.10. 

Thus, although hard to reproduce, might be something worth investigating.

**Comment by oscardssmith** [comment]
something is obviously going horrifically wrong here. It will be very hard to reproduce without an RR trace.

**Comment by mbauman** [comment]
Are you using `xi` in a `muladd` in your case?  In the linked CI runs, both are happening in [`cospi`](https://github.com/JuliaLang/julia/blob/6e5e87b2cafda840b90347c2e74202d2608d7c29/base/special/trig.jl#L839-L841) or [`sincospi`](https://github.com/JuliaLang/julia/blob/6e5e87b2cafda840b90347c2e74202d2608d7c29/base/special/trig.jl#L875-L877), where the nearby `muladd` looks suspiciously like it might result in something like https://github.com/JuliaLang/julia/issues/55785?

**Comment by oscardssmith** [comment]
the really confusing thing here is that the error message shows the number as being an int...

**Comment by lmiq** [comment]
The function where the error appeared for me was this one:

```julia
@inline function particle_cell(x::SVector{N}, box::Box) where {N}
    CartesianIndex(
        ntuple(N) do i
            xmin = box.computing_box[1][i]
            xi = (x[i] - xmin) / box.cell_size[i]
            index = floor(Int, xi) + 1
            return index
        end
    )
end
```

(from here: https://github.com/m3g/CellListMap.jl/blob/f85ede5118462331d23303833249f84ac6942ff3/src/Box.jl#L584)

**Comment by giordano** [comment]
Tangentially related, since as far as I understand `@testset` sets the RNG status internally, it'd be nice if `@testset` also printed the RNG status to let people reproduce the failure.

**Comment by lmiq** [comment]
Unfortunately I rerun the job and I can't reproduce it, but I'm using `StableRNGs`. So I'm not sure if that would really help.

**Comment by mbauman** [comment]
> the really confusing thing here is that the error message shows the number as being an int

Yes, that's very fast-mathy.  `Int` conversion does the following:

```llvm
    %1 = call double @llvm.trunc.f64(double %0)
    %2 = fsub double %0, %1
    %3 = fcmp une double %2, 0.000000e+00
    br i1 %3, label %error_path, label %ok_path
```

And then the `%error_path` uses `%0`.  If `%2` fuses into an fma with an inlined/prior multiplication against `%0`, you could get a different answer for the `br`.  But this should be reproducible with a stable RNG, so it may be a red herring.

**Comment by giordano** [comment]
That's not annotated `contract` though, is it?

**Comment by mbauman** [comment]
Correct. This could be dependent upon surrounding code if it inlines, but neither `sinpi` nor `sincospi` inline, so I don't think this is reason here.

**Comment by oscardssmith** [comment]
even if it is, that doesn't actually explain anything. This error can only occur if the compiler proves that `xi` is outside of `[-2e64, 2e64)` which `1.0` obviously isn't, so this bug requires a pretty dramatic miscompile.

**Comment by lmiq** [comment]
I got the error again in this workflow run:

**Comment by oscardssmith** [comment]
does RR work for Mac? I really want to see what is going on here

**Comment by lmiq** [comment]
I didn't know that existed. 

I logged into a macOS-latest machine, and ran hundreds of billions of `floor(Int, 1000*rand())` operations, and didn't get any error. Also, I ran a few times the tests of my package, also without any error. 

Unfortunately, this is really hard to reproduce. I thought I would get it failing easily, since I got the error often today when pushing new commits to the repo. 

I don't know if that helps, but this is the CI workflow of the repo where the issue appears: https://github.com/m3g/ComplexMixtures.jl/blob/main/.github/workflows/ci.yml

and this is the CI workflow of the tmate session I used lastly: https://github.com/lmiq/tmate/blob/main/.github/workflows/ci.yml

Maybe there is some configuration difference that someone can pinpoint.

**Comment by IanButterworth** [comment]
I've not reviewed this in detail but one of your failing CI runs was with coverage on and bounds checking forced on. You might need those to reproduce?

See the args that the packages were precompiled for, in the logs.

**Comment by IanButterworth** [comment]
For reproducing locally, julia-runtest runs now show how to rerun locally with the same config https://github.com/julia-actions/julia-runtest/pull/124

**Comment by lmiq** [comment]
I got now this weird error, which I guess is related to this one:

ˋˋˋ
ERROR: LoadError: TaskFailedException

    nested task error: DomainError with 2.0522995635873436:
    sin(x) is only defined for finite x.
    Stacktrace:
     [1] sin_domain_error(x::Float64)
ˋˋˋ

From: https://github.com/m3g/ComplexMixtures.jl/actions/runs/11864078437/job/33066804878#step:6:1790

(I keep getting the inexact Int error from time to time)

**Comment by lmiq** [comment]
I was never able to reproduce the bug myself in my machine or by logging into MacOS CI machine, so no. If I ever manage to do that, of course I will.

**Comment by giordano** [comment]
That is quite clear to me but I suggested to make your tests print the relevant variables to let people reproduce the issue.

**Comment by lmiq** [comment]
The original issue (the Inexact Int one) occurs in a ˋfloot(Int, x)ˋ instruction (where ˋxˋ is a float). Honestly I don´t know what I could potentially print to help. 

Also, this happens in a tight loop, and even adding a ˋtry' block to catch the error with more data would be hard, even for testing. I tried to simplify the issue all I could (for instance generating millions of floats to see if I catch that error for any), but with no avail.  

I´m not doing anything strange (not messing up with any unsafe operation, fastmath, or internals). So I am actually suprised if this is not occuring more frequently in the ecosystem. Maybe I should post this in other forums to see if anyone else is experiencing these errors.

**Comment by giordano** [comment]
> Also, this happens in a tight loop, and even adding a ˋtry' block to catch the error with more data would be hard, even for testing. I tried to simplify the issue all I could (for instance generating millions of floats to see if I catch that error for any), but with no avail.

But how have you tried? Did you use all the flags used for testing as suggested above, since they can affect compilation? And when the test fail you can print the input arguments to the test without touching the source code.

**Comment by lmiq** [comment]
As far as I remember (it has been some weeks) I did all that. Logged into the CI machine and followed all those instructions. Ran the tests a dozen of times and tried to run a simple `floor` on millions of random floats to see if the error occurred. 

I'll try again when I have some time.

**Comment by IanButterworth** [comment]
I noticed that I think all the CI runs here are for macOS x64 on macos-latest which for some time now has resolved to `macos-14-arm64` so these are happening in x64 julia runs via rosetta on apple silicon machines.

Could it be a rosetta bug?

**Comment by LilithHafner** [comment]
I ran into this with `InexactError: Int64(0.0)` here: https://github.com/LilithHafner/Chairmarks.jl/pull/138#issuecomment-2496015782

CI logs here: https://github.com/LilithHafner/Chairmarks.jl/actions/runs/11996296487/job/33440567741

The lowest line in the stack trace that is in my package is https://github.com/LilithHafner/Chairmarks.jl/blob/ea0fe664836d677823fa784d45702cf15702ac4f/src/benchmarking.jl#L93

```julia
max(1, floor(Int, target_sample_time/(something(calibration2time, calibration1time)+1e-9)))
```

This line has been run millions of times (maybe billions) and this is the first time it's thrown, that I know of. I have not been able to reproduce it. I believe `target_sample_time` was `2.8717779972764167e-5` at the time this threw but don't have a record of `calibration2time` or `calibration1time`. They should both be a positive Float64 or nothing.

I don't pirate conversion or anything like that. This is Julia 1.11.1, MacOS-latest.

The test logs include
```
Precompiling project for configuration --code-coverage=@/Users/runner/work/Chairmarks.jl/Chairmarks.jl --color=yes --check-bounds=yes --warn-overwrite=yes --depwarn=yes --inline=yes --startup-file=no --track-allocation=none --check-bounds=yes --compiled-modules=yes...
```
So those are probably the julia flags.

Smells like a kernel bug to me, though I can't be sure.

**Comment by IanButterworth** [comment]
Are you intentionally running x64 on Apple Silicon runners? i.e. https://github.com/JuliaLang/julia/issues/55878#issuecomment-2480734673

If not, there is now a warning on https://github.com/LilithHafner/Chairmarks.jl/actions/runs/11996296487 but I see it's one of many..

**Comment by LilithHafner** [comment]
No, not intentionally. https://github.com/LilithHafner/Chairmarks.jl/issues/139.

**Comment by lmiq** [comment]
I think the issue is clear now, and a solution is available (or at least under way, with the warning). I updated my CI files and I'm not getting the error anymore. 

Feel free to mark the issue as closed whenever you think it is appropriate.

Thanks.

**Comment by IanButterworth** [comment]
Maybe we should say that x86 on macOS aarch64 is unmaintained/unsupported though. It's not like Base tests are run for that combination.

**Comment by LilithHafner** [comment]
We already don't support x86 on any mac. Do you mean we should also drop x86_64 on aarch64? The Intel Mac Pro was discontinued in June 2023 and it only supports x86_64. Do we run Base tests for that hardware?
